### ISSUE TITLE
Introduce BeforeTestInstantiationCallback extension API

### ISSUE THREAD

**Issue by codecholeric** [junit-team/junit5#839]
## Short story:

For our test infrastructure to create test data, we must distinguish calls during the static initialization of each test class from calls during test instantiation for each test method.
This could be solved by the proposed `BeforeTestInstantiationCallback`.

## Long story:

When we write our tests, we want to restrict the data that have to be set explicitly to those that the specific test requires. All other properties are set randomly, but valid. Naturally using random data may introduce randomly failing tests, so we need the means to easily reproduce test failures that e.g. happen in the CI environment.
Reproducing such failures requires knowing the seed values for our random data when the specific test was executed. This is where we want to take advantage of a JUnit 5 extension. It has to store one seed value at the class level, for example if static constants are initialized using our random mechanism. In addition, it has to store the seed value at the instance level, for each test run. If a test should fail, the respective seed values must be reported in the failure message. If the failure is related to the use of random data, all this together allows us to reproduce it in any other environment. Example:

```java
private static final Address staticAddress = 
    createRandomAddress(); // determined by the class seed
 
private Address instanceAddress = 
    createRandomAddressBuilder() // determined by the instance seed
        .withStreet("some street")
        .build();
 
@Test
void someTest() {
    Address anotherAddress = 
        createRandomAddress(); // also determined by the instance seed
    // ...
}
```

We note the class seed in a `BeforeAllCallback`, which is the perfect place. The switch to using the instance seed however, has to be right before the test instance is created, to cover random data used for instance field initialization, which is a rather common use case in our tests.
Obviously, the `TestInstancePostProcessor` callback is too late for switching from the class seed to the instance seed. Currently we have an ugly workaround, where we check the stacktrace to switch to the instance seed, as soon as we detect a call from the test's constructor. If there are no such calls we switch to the instance seed in the `TestInstancePostProcessor` callback.

The proposed `BeforeTestInstantiationCallback` would be the natural place to make the required switch to the instance seed trivial.

Do you think, adding the `BeforeTestInstantiationCallback` would be harmful?

**Comment by nicolaiparlog** [comment]
Thanks for opening the issue! There are some details I need clarifications on because I'm not quite getting what you describe. (Excuse me if these are dumb question. :smiley:)

> It has to store one seed value at the class level, for example if static constants are initialized using our random mechanism.  [... much later ...] We note the class seed in a `BeforeAllCallback`, which is the perfect place.

What exactly do you mean by "note"? Since `staticAddress` is initialized as part of the class' initialization, the `Random` instance must also already have been constructed before any `BeforeAllCallback` is called. How can your extension then access the seed used to create that `Random`? Is it stored somewhere?

> The switch to using the instance seed however, has to be right before the test instance is created

What do you mean with "switch"? Do you create a new `Random` instance when the first instance field is initialized? If so, how? Or would you want to make that switch but the current extension model does not allow you to?

Another, more general question: Obviously the fields must always be initialized in the same order for the randomization to be reproducible. Does the JLS or JVMLS guarantee an order? Because if not, the entire premise is built on an unspecified implementation detail.

**Comment by codecholeric** [comment]
Sorry, that it took me a while to answer!! Thanks for taking the time to look into this :smiley: I'm not amazed, that I did not get across clearly, it's not the easiest thing to describe somehow :wink:

By "note" I just meant to set and store the current seed somewhere for later (set and store static seed -> static init -> set and store instance seed -> instance init)

The `BeforeAllCallback` is executed before the class initialization, so it is early enough to set the 'static seed' for the class initialization block. If a `BeforeAllCallback` sets the static seed, then `staticAddress` as part of the class init, will be able to use it. In case that is an internal detail of JUnit 5 that may change in time, we might have to introduce some sort of lazy static seed creation on first access (similar to our workaround for the instance seed problem, but hopefully then less shady).

Yes, at the moment right before any instance related initialization first happens (field init, constructor, ...) we want to create a new `Random`, where the seed is derived from the static `Random`, so we can make the whole process deterministic. However this is where we are missing an extension point, that is called right before any `testClass.newInstance()` is called by JUnit.

As far as I understand http://docs.oracle.com/javase/specs/jls/se8/html/jls-12.html#jls-12.5 the order should be deterministic, "in the left-to-right order in which they appear textually in the source code", so theoretically (and empirically so far :wink:) this should work.

**Comment by nicolaiparlog** [comment]
Thanks for clearing this up, this is an interesting problem. :smiley:

> Do you think, adding the `BeforeTestInstantiationCallback` would be harmful?

Forgot to answer that... No, not at all. It's of course up to people with more knowledge of the internals (summoning @sormuras) to decide that but I see no conceptual reason why that would be a bad idea.

Still, gnawing on your problem as it is... Did you ever consider injecting `Random` into setup methods? I created a little [proof of concept](https://gist.github.com/nicolaiparlog/b1ebad57dbb0a15a32283e9b741ec437) that injects an instance of `Random` into setup and test methods and also catches exceptions to print the seed of the failed test. You would then use the setup methods to populate whatever fields you need.

One caveat is that reflection order is not guaranteed, so injecting the same `Random` instance into multiple setup methods might not be the best idea. The extension could of course check that and emit a warning...

**Comment by sbrannen** [comment]
@codecholeric,

> The `BeforeAllCallback` is executed before the class initialization, so it is early enough to set the 'static seed' for the class initialization block. If a `BeforeAllCallback` sets the static seed, then `staticAddress` as part of the class init, will be able to use it. In case that is an internal detail of JUnit 5 that may change in time, we might have to introduce some sort of lazy static seed creation on first access (similar to our workaround for the instance seed problem, but hopefully then less shady).

When you say that the "`BeforeAllCallback` is executed before the class initialization," are you referring to a _static initialization block_? If so, then I don't see how that could be true since JUnit must have access to a `Class` object for the test class before invoking any `BeforeAllCallback`. Hence, the class would already have been loaded in the `ClassLoader` by that point, and any static initialization blocks would already have been executed.

Please note as well that if a user annotates a test class with `@TestInstance(Lifecycle.PER_CLASS)` (feature introduced in 5.0 M5)... then JUnit will instantiate the test class before invoking any `BeforeAllCallback`. 😉

**Comment by codecholeric** [comment]
Yes, I was referring to a static initialization block, and I think JUnit 5 at the moment does sth. like `Class.forName(testClassName, false, classLoader)`, because the static block is indeed (at least with my old example using M3) evaluated after the callback. E.g.

```
public class FooBeforeAll implements BeforeAllCallback {
    @Override
    public void beforeAll(ContainerExtensionContext context) throws Exception {
        System.out.println("BEFORE ALL");
    }
}
```

```
@ExtendWith(FooBeforeAll.class)
public class FooTest {
    static {
        System.out.println("STATIC");
    }

    @Test
    void name() {
        System.out.println("TESTS");
    }
}
```

will print

```
BEFORE ALL
STATIC
TESTS
```

But anyway, I'm not convinced, that it is a good idea, to rely on that :wink: 

So, we would have to add some sort of lazy mechanism for this, too.
I think if I would start from scratch, I would follow @nicolaiparlog 's approach, and kick all this evil static factory code (even though it causes a little more boiler plate to set all those fields, but in my opinion we could just forbid static random fields in tests anyway).
However, the other question is, if it is not a valid request in general, to have an extension point right before a test class is instantiated.

**Comment by sbrannen** [comment]
> But anyway, I'm not convinced, that it is a good idea, to rely on that 😉

I agree: that's fragile at best.

In your example it's obviously true that static initialization blocks of the test class are not invoked prior to invocation of `BeforeAllCallback` callbacks. That's because there is no "active use" of the class prior to that.

However, the following (a bit contrived) example demonstrates that that will not always be the case.

```java
@ExtendWith(FooBeforeAll.class)
class FooTest {

	static FooTest INSTANCE = new FooTest();

	static {
		System.out.println("STATIC");
	}

	@org.junit.jupiter.api.Test
	void name() {
		System.out.println("TESTS");
	}

}

class FooBeforeAll implements BeforeAllCallback {

	@Override
	public void beforeAll(ExtensionContext context) throws Exception {
		System.out.println("BEFORE ALL: " + FooTest.INSTANCE.getClass().getSimpleName());
	}

}
```

If any other class _actively uses_ the test class earlier, the output to the console will be the following.

```
STATIC
BEFORE ALL: FooTest
TESTS
```

So, in summary, I would not implement an extension that _hopes_ that the test class has not yet been fully initialized, because as a wise person once said, "Hope is not a strategy!". 😉

**Comment by codecholeric** [comment]
Okay, I guess this concrete example makes it even more clear, that relying on the order of static vs beforeAll is a bad idea :wink: 
I guess the lazy mechanism is the only choice then...

**Comment by sbrannen** [comment]
> However, the other question is, if it is not a valid request in general, to have an extension point right before a test class is instantiated.

_We're still pondering that._

In other words, we're basically waiting to see if the community at large really needs such a feature (i.e., whether it warrants inclusion in the extension API).

**Comment by jonashoef** [comment]
Hi again, 
with JUnit 5.5, you provided the [InvocationInterceptor](https://junit.org/junit5/docs/5.5.2/api/org/junit/jupiter/api/extension/InvocationInterceptor.html), which solves our problem elegantly via the interceptTestClassConstructor​()-method. 

Good job :-), definitely a much more powerful and general enhancement than our original proposal. 

One thing that I wasn't aware of in that context is the fact that the test class may not have been initialized yet (clinit), when the interceptTestClassConstructor​()-method is invoked. It may be worth mentioning that in the javadoc of the method.

**Comment by sbrannen** [comment]
@jonashoef, can you please create a new issue (or even a PR) for that?

Otherwise comments like these get lost the backlog.

Thanks

**Comment by jonashoef** [comment]
Provided [PR](https://github.com/junit-team/junit5/pull/2103)
